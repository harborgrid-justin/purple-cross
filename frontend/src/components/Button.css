/* Button Component Styles */

.btn {
  font-family: inherit;
  font-weight: 500;
  border: none;
  border-radius: var(--radius-md, 0.5rem);
  cursor: pointer;
  transition: all 0.2s ease-in-out;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--spacing-xs, 0.25rem);
  text-decoration: none;
  line-height: 1.5;
}

.btn:focus-visible {
  outline: 2px solid var(--primary-color, #7c3aed);
  outline-offset: 2px;
}

.btn:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

/* Button Sizes */
.btn-sm {
  padding: var(--spacing-xs, 0.25rem) var(--spacing-sm, 0.5rem);
  font-size: 0.875rem;
  min-height: 32px;
}

.btn-md {
  padding: var(--spacing-sm, 0.5rem) var(--spacing-lg, 1.5rem);
  font-size: 1rem;
  min-height: 44px; /* WCAG 2.1 AA touch target */
}

.btn-lg {
  padding: var(--spacing-md, 1rem) var(--spacing-xl, 2rem);
  font-size: 1.125rem;
  min-height: 48px;
}

/* Button Variants */
.btn-primary {
  background-color: var(--primary-color, #7c3aed);
  color: white;
}

.btn-primary:hover:not(:disabled) {
  background-color: var(--primary-dark, #6d28d9);
}

.btn-primary:active:not(:disabled) {
  background-color: #5b21b6;
}

.btn-secondary {
  background-color: var(--bg-primary, #ffffff);
  color: var(--text-primary, #1f2937);
  border: 1px solid var(--border-color, #e5e7eb);
}

.btn-secondary:hover:not(:disabled) {
  background-color: var(--bg-secondary, #f9fafb);
  border-color: var(--text-secondary, #6b7280);
}

.btn-danger {
  background-color: var(--danger-color, #ef4444);
  color: white;
}

.btn-danger:hover:not(:disabled) {
  background-color: #dc2626;
}

.btn-danger:active:not(:disabled) {
  background-color: #b91c1c;
}

.btn-success {
  background-color: var(--secondary-color, #10b981);
  color: white;
}

.btn-success:hover:not(:disabled) {
  background-color: #059669;
}

.btn-success:active:not(:disabled) {
  background-color: #047857;
}

/* Button Modifiers */
.btn-full-width {
  width: 100%;
}

.btn-loading {
  opacity: 0.7;
  cursor: wait;
  pointer-events: none;
}

/* Spinner Animation */
.spinner {
  width: 1em;
  height: 1em;
  border: 2px solid currentColor;
  border-top-color: transparent;
  border-radius: 50%;
  animation: spin 0.6s linear infinite;
}

@keyframes spin {
  to {
    transform: rotate(360deg);
  }
}

/* Reduced Motion Support */
@media (prefers-reduced-motion: reduce) {
  .spinner {
    animation-duration: 1.2s;
  }
}
