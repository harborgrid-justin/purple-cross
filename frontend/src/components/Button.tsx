/**
 * WF-COMP-XXX | Button.tsx - Button
 * Purpose: React component for Button functionality
 * Dependencies: react
 * Last Updated: 2025-10-23 | File Type: .tsx
 */

import { ButtonHTMLAttributes, ReactNode } from 'react';
import clsx from 'clsx';

/**
 * Props for the Button component.
 *
 * @interface ButtonProps
 * @extends {ButtonHTMLAttributes<HTMLButtonElement>}
 *
 * @property {ReactNode} children - The content to display inside the button
 * @property {'primary' | 'secondary' | 'danger' | 'success'} [variant='primary'] - Visual style variant
 * @property {'sm' | 'md' | 'lg'} [size='md'] - Button size
 * @property {boolean} [fullWidth=false] - Whether button should take full width of container
 * @property {boolean} [loading=false] - Loading state - shows spinner and disables button
 *
 * @example
 * // Primary button
 * <Button variant="primary" onClick={handleClick}>
 *   Save Changes
 * </Button>
 *
 * @example
 * // Loading state
 * <Button loading={isSubmitting}>
 *   Submit Form
 * </Button>
 */
interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode;
  variant?: 'primary' | 'secondary' | 'danger' | 'success';
  size?: 'sm' | 'md' | 'lg';
  fullWidth?: boolean;
  loading?: boolean;
}

/**
 * Button component with multiple variants, sizes, and loading state support.
 *
 * @component
 * @param {ButtonProps} props - The component props
 * @returns {JSX.Element} Rendered button element
 *
 * @example
 * // Basic usage
 * <Button onClick={handleSave}>Save</Button>
 *
 * @example
 * // With variant and size
 * <Button variant="danger" size="lg" onClick={handleDelete}>
 *   Delete Account
 * </Button>
 *
 * @example
 * // Full width button
 * <Button fullWidth variant="primary" onClick={handleSubmit}>
 *   Submit Form
 * </Button>
 *
 * @example
 * // Loading state
 * <Button loading={isProcessing} disabled={!isValid}>
 *   Process Payment
 * </Button>
 *
 * @remarks
 * - Extends all native button HTML attributes
 * - Automatically disables button when loading={true}
 * - Uses clsx for conditional className composition
 * - Loading state shows spinner with "Loading..." text
 * - Supports custom className for additional styling
 */
const Button = ({
  children,
  variant = 'primary',
  size = 'md',
  fullWidth = false,
  loading = false,
  disabled,
  className,
  ...props
}: ButtonProps) => {
  return (
    <button
      className={clsx(
        'btn',
        `btn-${variant}`,
        `btn-${size}`,
        {
          'btn-full-width': fullWidth,
          'btn-loading': loading,
        },
        className
      )}
      disabled={disabled || loading}
      {...props}
    >
      {loading ? (
        <>
          <span className="spinner" aria-hidden="true"></span>
          <span>Loading...</span>
        </>
      ) : (
        children
      )}
    </button>
  );
};

export default Button;
