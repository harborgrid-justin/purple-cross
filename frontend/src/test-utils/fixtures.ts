/**
 * WF-COMP-XXX | fixtures.ts - Test Fixtures
 * Purpose: Mock data fixtures for testing
 * Dependencies: None
 * Last Updated: 2025-10-24 | File Type: .ts
 */

/**
 * Mock patient data
 */
export const mockPatient = {
  id: '1',
  name: 'Buddy',
  species: 'Dog',
  breed: 'Golden Retriever',
  dateOfBirth: '2020-01-15',
  gender: 'Male',
  color: 'Golden',
  weight: 32.5,
  microchipNumber: '123456789012345',
  status: 'active',
  clientId: '1',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockPatients = [
  mockPatient,
  {
    id: '2',
    name: 'Whiskers',
    species: 'Cat',
    breed: 'Siamese',
    dateOfBirth: '2019-05-20',
    gender: 'Female',
    color: 'Cream',
    weight: 4.2,
    microchipNumber: '987654321098765',
    status: 'active',
    clientId: '1',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
  {
    id: '3',
    name: 'Max',
    species: 'Dog',
    breed: 'German Shepherd',
    dateOfBirth: '2018-03-10',
    gender: 'Male',
    color: 'Black and Tan',
    weight: 38.0,
    status: 'inactive',
    clientId: '2',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock client data
 */
export const mockClient = {
  id: '1',
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@example.com',
  phone: '555-0100',
  address: '123 Main St',
  city: 'Springfield',
  state: 'IL',
  zipCode: '62701',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockClients = [
  mockClient,
  {
    id: '2',
    firstName: 'Jane',
    lastName: 'Smith',
    email: 'jane.smith@example.com',
    phone: '555-0200',
    address: '456 Oak Ave',
    city: 'Springfield',
    state: 'IL',
    zipCode: '62702',
    status: 'active',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock appointment data
 */
export const mockAppointment = {
  id: '1',
  patientId: '1',
  clientId: '1',
  staffId: '1',
  appointmentType: 'checkup',
  startTime: '2025-10-25T10:00:00Z',
  endTime: '2025-10-25T10:30:00Z',
  status: 'scheduled',
  notes: 'Annual wellness exam',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockAppointments = [
  mockAppointment,
  {
    id: '2',
    patientId: '2',
    clientId: '1',
    staffId: '1',
    appointmentType: 'surgery',
    startTime: '2025-10-25T14:00:00Z',
    endTime: '2025-10-25T15:00:00Z',
    status: 'confirmed',
    notes: 'Spay procedure',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock medical record data
 */
export const mockMedicalRecord = {
  id: '1',
  patientId: '1',
  staffId: '1',
  visitDate: '2025-10-20',
  chiefComplaint: 'Annual checkup',
  diagnosis: 'Healthy',
  treatment: 'Vaccinations administered',
  prescriptions: [],
  labTests: [],
  status: 'completed',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockMedicalRecords = [
  mockMedicalRecord,
  {
    id: '2',
    patientId: '1',
    staffId: '1',
    visitDate: '2025-09-15',
    chiefComplaint: 'Ear infection',
    diagnosis: 'Otitis externa',
    treatment: 'Antibiotic ear drops',
    prescriptions: ['1'],
    labTests: [],
    status: 'completed',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock prescription data
 */
export const mockPrescription = {
  id: '1',
  patientId: '1',
  medicationName: 'Amoxicillin',
  dosage: '250mg',
  frequency: 'Twice daily',
  duration: '10 days',
  instructions: 'Give with food',
  prescribedBy: '1',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock staff data
 */
export const mockStaff = {
  id: '1',
  firstName: 'Dr. Sarah',
  lastName: 'Johnson',
  email: 'sarah.johnson@purplecross.com',
  phone: '555-0300',
  role: 'veterinarian',
  specialty: 'General Practice',
  licenseNumber: 'VET-12345',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock invoice data
 */
export const mockInvoice = {
  id: '1',
  clientId: '1',
  patientId: '1',
  invoiceNumber: 'INV-2025-001',
  invoiceDate: '2025-10-20',
  dueDate: '2025-11-20',
  subtotal: 150.0,
  tax: 12.0,
  total: 162.0,
  amountPaid: 0.0,
  status: 'unpaid',
  items: [
    {
      description: 'Annual wellness exam',
      quantity: 1,
      unitPrice: 100.0,
      total: 100.0,
    },
    {
      description: 'Rabies vaccination',
      quantity: 1,
      unitPrice: 50.0,
      total: 50.0,
    },
  ],
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock inventory item data
 */
export const mockInventoryItem = {
  id: '1',
  name: 'Amoxicillin 250mg',
  category: 'Medication',
  sku: 'MED-001',
  quantity: 100,
  reorderLevel: 20,
  unitPrice: 2.5,
  supplier: 'VetSupply Co',
  expirationDate: '2026-12-31',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock lab test data
 */
export const mockLabTest = {
  id: '1',
  patientId: '1',
  testType: 'Blood panel',
  testDate: '2025-10-20',
  results: 'All values within normal range',
  status: 'completed',
  priority: 'routine',
  orderedBy: '1',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock document data
 */
export const mockDocument = {
  id: '1',
  patientId: '1',
  clientId: '1',
  title: 'Vaccination Certificate',
  documentType: 'Certificate',
  fileName: 'vaccination-cert.pdf',
  fileSize: 245678,
  mimeType: 'application/pdf',
  uploadedBy: '1',
  status: 'signed',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock user data
 */
export const mockUser = {
  id: '1',
  email: 'test@example.com',
  firstName: 'Test',
  lastName: 'User',
  role: 'veterinarian',
  token: 'mock-jwt-token',
};

/**
 * Mock API response
 */
export const mockApiResponse = <T>(data: T) => ({
  status: 'success',
  data,
});

/**
 * Mock API error response
 */
export const mockApiError = (message: string, statusCode = 400) => ({
  status: 'error',
  message,
  statusCode,
});

/**
 * Mock pagination metadata
 */
export const mockPagination = {
  page: 1,
  limit: 10,
  total: 100,
  totalPages: 10,
};

/**
 * Generate mock list with pagination
 */
export const mockPaginatedResponse = <T>(
  data: T[],
  page = 1,
  limit = 10,
  total?: number
) => ({
  status: 'success',
  data,
  pagination: {
    page,
    limit,
    total: total || data.length,
    totalPages: Math.ceil((total || data.length) / limit),
  },
});
