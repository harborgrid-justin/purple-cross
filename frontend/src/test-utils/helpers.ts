/**
 * WF-COMP-XXX | helpers.ts - Test Helper Functions
 * Purpose: Common utilities for testing
 * Dependencies: @testing-library/react
 * Last Updated: 2025-10-24 | File Type: .ts
 */

import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Create a user event instance for simulating user interactions
 * This should be called at the beginning of each test
 *
 * @example
 * ```tsx
 * const user = createUser();
 * await user.click(button);
 * await user.type(input, 'text');
 * ```
 */
export const createUser = () => {
  return userEvent.setup();
};

/**
 * Wait for an async operation to complete with custom timeout
 */
export const waitForAsync = async (
  callback: () => void | Promise<void>,
  timeout = 3000
): Promise<void> => {
  await waitFor(callback, { timeout });
};

/**
 * Wait for loading states to complete
 * Useful for components that show loading spinners
 */
export const waitForLoadingToFinish = async (
  getByText: (text: string | RegExp) => HTMLElement,
  timeout = 3000
): Promise<void> => {
  await waitFor(
    () => {
      expect(getByText(/loading/i)).not.toBeInTheDocument();
    },
    { timeout }
  );
};

/**
 * Mock window.scrollTo (not implemented in jsdom)
 */
export const mockScrollTo = (): void => {
  window.scrollTo = vi.fn();
};

/**
 * Mock localStorage for testing
 */
export const mockLocalStorage = (): {
  getItem: ReturnType<typeof vi.fn>;
  setItem: ReturnType<typeof vi.fn>;
  removeItem: ReturnType<typeof vi.fn>;
  clear: ReturnType<typeof vi.fn>;
} => {
  const storage: Record<string, string> = {};

  return {
    getItem: vi.fn((key: string) => storage[key] || null),
    setItem: vi.fn((key: string, value: string) => {
      storage[key] = value;
    }),
    removeItem: vi.fn((key: string) => {
      delete storage[key];
    }),
    clear: vi.fn(() => {
      Object.keys(storage).forEach((key) => delete storage[key]);
    }),
  };
};

/**
 * Mock sessionStorage for testing
 */
export const mockSessionStorage = mockLocalStorage;

/**
 * Delay execution for a specified time (in milliseconds)
 * Useful for testing loading states or debounced functions
 */
export const delay = (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

/**
 * Create a mock file for file upload testing
 */
export const createMockFile = (
  name: string,
  content: string,
  type: string
): File => {
  const blob = new Blob([content], { type });
  return new File([blob], name, { type });
};

/**
 * Create a mock image file for image upload testing
 */
export const createMockImageFile = (
  name = 'test-image.png',
  size = 1024
): File => {
  const buffer = new ArrayBuffer(size);
  const blob = new Blob([buffer], { type: 'image/png' });
  return new File([blob], name, { type: 'image/png' });
};

/**
 * Assert that an element has a specific class
 */
export const expectToHaveClass = (
  element: HTMLElement,
  className: string
): void => {
  expect(element.classList.contains(className)).toBe(true);
};

/**
 * Assert that an element does not have a specific class
 */
export const expectNotToHaveClass = (
  element: HTMLElement,
  className: string
): void => {
  expect(element.classList.contains(className)).toBe(false);
};

/**
 * Get computed style of an element
 */
export const getComputedStyleValue = (
  element: HTMLElement,
  property: string
): string => {
  return window.getComputedStyle(element).getPropertyValue(property);
};

/**
 * Simulate a network delay for testing loading states
 */
export const simulateNetworkDelay = async (ms = 500): Promise<void> => {
  await delay(ms);
};

/**
 * Create a spy for console methods (useful for testing error handling)
 */
export const spyOnConsole = (): {
  log: ReturnType<typeof vi.spyOn>;
  warn: ReturnType<typeof vi.spyOn>;
  error: ReturnType<typeof vi.spyOn>;
} => {
  return {
    log: vi.spyOn(console, 'log').mockImplementation(() => {}),
    warn: vi.spyOn(console, 'warn').mockImplementation(() => {}),
    error: vi.spyOn(console, 'error').mockImplementation(() => {}),
  };
};

/**
 * Assert that a function was called with specific arguments
 */
export const expectCalledWith = <T extends (...args: any[]) => any>(
  fn: T,
  ...args: Parameters<T>
): void => {
  expect(fn).toHaveBeenCalledWith(...args);
};

/**
 * Wait for a specific number of milliseconds in tests
 * Uses vi.advanceTimersByTime if fake timers are enabled
 */
export const advanceTime = async (ms: number): Promise<void> => {
  if (vi.isFakeTimers()) {
    vi.advanceTimersByTime(ms);
  } else {
    await delay(ms);
  }
};

/**
 * Type assertion helper for better type inference in tests
 */
export const assertType = <T>(_value: T): void => {
  // Type assertion helper - no runtime behavior
};
