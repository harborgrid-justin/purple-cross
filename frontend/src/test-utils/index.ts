/**
 * WF-COMP-XXX | index.ts - Test Utils Exports
 * Purpose: Central export point for all test utilities
 * Dependencies: None
 * Last Updated: 2025-10-24 | File Type: .ts
 */

// Export custom render functions
export * from './render';

// Export test helpers
export * from './helpers';

// Export test fixtures
export * from './fixtures';

// Export MSW mocks
export * from './mocks/handlers';
export { server, mockServer } from './mocks/server';
