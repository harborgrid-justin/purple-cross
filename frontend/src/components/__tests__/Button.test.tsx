/**
 * WF-COMP-XXX | Button.test.tsx - Button Component Tests
 * Purpose: Comprehensive tests for Button component
 * Dependencies: @testing-library/react, @testing-library/user-event
 * Last Updated: 2025-10-24 | File Type: .tsx
 */

import { describe, it, expect, vi } from 'vitest';
import { renderWithoutRouter, createUser } from '@/test-utils';
import Button from '../Button';

describe('Button Component', () => {
  describe('Rendering', () => {
    it('should render with default props', () => {
      const { getByRole } = renderWithoutRouter(<Button>Click me</Button>);

      const button = getByRole('button', { name: /click me/i });
      expect(button).toBeInTheDocument();
      expect(button).toHaveClass('btn');
      expect(button).toHaveClass('btn-primary');
      expect(button).toHaveClass('btn-md');
    });

    it('should render children correctly', () => {
      const { getByText } = renderWithoutRouter(<Button>Save Changes</Button>);

      expect(getByText('Save Changes')).toBeInTheDocument();
    });

    it('should render with custom className', () => {
      const { getByRole } = renderWithoutRouter(
        <Button className="custom-class">Click me</Button>
      );

      const button = getByRole('button');
      expect(button).toHaveClass('custom-class');
      expect(button).toHaveClass('btn'); // Should still have base class
    });
  });

  describe('Variants', () => {
    it('should render primary variant', () => {
      const { getByRole } = renderWithoutRouter(
        <Button variant="primary">Primary</Button>
      );

      expect(getByRole('button')).toHaveClass('btn-primary');
    });

    it('should render secondary variant', () => {
      const { getByRole } = renderWithoutRouter(
        <Button variant="secondary">Secondary</Button>
      );

      expect(getByRole('button')).toHaveClass('btn-secondary');
    });

    it('should render danger variant', () => {
      const { getByRole } = renderWithoutRouter(
        <Button variant="danger">Delete</Button>
      );

      expect(getByRole('button')).toHaveClass('btn-danger');
    });

    it('should render success variant', () => {
      const { getByRole } = renderWithoutRouter(
        <Button variant="success">Success</Button>
      );

      expect(getByRole('button')).toHaveClass('btn-success');
    });
  });

  describe('Sizes', () => {
    it('should render small size', () => {
      const { getByRole } = renderWithoutRouter(<Button size="sm">Small</Button>);

      expect(getByRole('button')).toHaveClass('btn-sm');
    });

    it('should render medium size (default)', () => {
      const { getByRole } = renderWithoutRouter(<Button size="md">Medium</Button>);

      expect(getByRole('button')).toHaveClass('btn-md');
    });

    it('should render large size', () => {
      const { getByRole } = renderWithoutRouter(<Button size="lg">Large</Button>);

      expect(getByRole('button')).toHaveClass('btn-lg');
    });
  });

  describe('Full Width', () => {
    it('should render full width button', () => {
      const { getByRole } = renderWithoutRouter(
        <Button fullWidth>Full Width</Button>
      );

      expect(getByRole('button')).toHaveClass('btn-full-width');
    });

    it('should not have full width class by default', () => {
      const { getByRole } = renderWithoutRouter(<Button>Normal</Button>);

      expect(getByRole('button')).not.toHaveClass('btn-full-width');
    });
  });

  describe('Loading State', () => {
    it('should render loading state', () => {
      const { getByRole, getByText } = renderWithoutRouter(
        <Button loading>Submit</Button>
      );

      const button = getByRole('button');
      expect(button).toHaveClass('btn-loading');
      expect(getByText('Loading...')).toBeInTheDocument();
      expect(button).toBeDisabled();
    });

    it('should show spinner in loading state', () => {
      const { container } = renderWithoutRouter(<Button loading>Submit</Button>);

      const spinner = container.querySelector('.spinner');
      expect(spinner).toBeInTheDocument();
      expect(spinner).toHaveAttribute('aria-hidden', 'true');
    });

    it('should not show children text when loading', () => {
      const { queryByText } = renderWithoutRouter(
        <Button loading>Submit Form</Button>
      );

      // Original button text should not be visible
      expect(queryByText('Submit Form')).not.toBeInTheDocument();
      // But Loading... text should be visible
      expect(queryByText('Loading...')).toBeInTheDocument();
    });
  });

  describe('Disabled State', () => {
    it('should render disabled button', () => {
      const { getByRole } = renderWithoutRouter(<Button disabled>Disabled</Button>);

      expect(getByRole('button')).toBeDisabled();
    });

    it('should be disabled when loading', () => {
      const { getByRole } = renderWithoutRouter(<Button loading>Loading</Button>);

      expect(getByRole('button')).toBeDisabled();
    });

    it('should not call onClick when disabled', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button disabled onClick={handleClick}>
          Click me
        </Button>
      );

      await user.click(getByRole('button'));
      expect(handleClick).not.toHaveBeenCalled();
    });

    it('should not call onClick when loading', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button loading onClick={handleClick}>
          Click me
        </Button>
      );

      await user.click(getByRole('button'));
      expect(handleClick).not.toHaveBeenCalled();
    });
  });

  describe('User Interactions', () => {
    it('should call onClick when clicked', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button onClick={handleClick}>Click me</Button>
      );

      await user.click(getByRole('button'));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('should call onClick multiple times', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button onClick={handleClick}>Click me</Button>
      );

      const button = getByRole('button');
      await user.click(button);
      await user.click(button);
      await user.click(button);

      expect(handleClick).toHaveBeenCalledTimes(3);
    });

    it('should receive event object in onClick', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button onClick={handleClick}>Click me</Button>
      );

      await user.click(getByRole('button'));
      expect(handleClick).toHaveBeenCalledWith(
        expect.objectContaining({
          type: 'click',
        })
      );
    });

    it('should support keyboard navigation (Enter key)', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button onClick={handleClick}>Click me</Button>
      );

      const button = getByRole('button');
      button.focus();
      await user.keyboard('{Enter}');

      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('should support keyboard navigation (Space key)', async () => {
      const handleClick = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Button onClick={handleClick}>Click me</Button>
      );

      const button = getByRole('button');
      button.focus();
      await user.keyboard(' ');

      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('HTML Attributes', () => {
    it('should forward type attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Button type="submit">Submit</Button>
      );

      expect(getByRole('button')).toHaveAttribute('type', 'submit');
    });

    it('should have button element', () => {
      const { getByRole } = renderWithoutRouter(<Button>Click</Button>);

      // Button should render as button element
      const button = getByRole('button');
      expect(button.tagName).toBe('BUTTON');
    });

    it('should forward aria-label', () => {
      const { getByRole } = renderWithoutRouter(
        <Button aria-label="Close dialog">×</Button>
      );

      expect(getByRole('button')).toHaveAttribute('aria-label', 'Close dialog');
    });

    it('should forward data attributes', () => {
      const { getByRole } = renderWithoutRouter(
        <Button data-testid="custom-button" data-action="save">
          Save
        </Button>
      );

      const button = getByRole('button');
      expect(button).toHaveAttribute('data-testid', 'custom-button');
      expect(button).toHaveAttribute('data-action', 'save');
    });

    it('should forward id attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Button id="submit-button">Submit</Button>
      );

      expect(getByRole('button')).toHaveAttribute('id', 'submit-button');
    });

    it('should forward name attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Button name="action">Action</Button>
      );

      expect(getByRole('button')).toHaveAttribute('name', 'action');
    });
  });

  describe('Accessibility', () => {
    it('should be keyboard accessible', () => {
      const { getByRole } = renderWithoutRouter(<Button>Click me</Button>);

      const button = getByRole('button');
      expect(button).toBeInTheDocument();
      expect(button.tagName).toBe('BUTTON');
    });

    it('should have proper role', () => {
      const { getByRole } = renderWithoutRouter(<Button>Click me</Button>);

      expect(getByRole('button')).toBeInTheDocument();
    });

    it('should support aria-disabled when disabled', () => {
      const { getByRole } = renderWithoutRouter(<Button disabled>Disabled</Button>);

      expect(getByRole('button')).toHaveAttribute('disabled');
    });

    it('should have aria-hidden on spinner', () => {
      const { container } = renderWithoutRouter(<Button loading>Loading</Button>);

      const spinner = container.querySelector('.spinner');
      expect(spinner).toHaveAttribute('aria-hidden', 'true');
    });
  });

  describe('Edge Cases', () => {
    it('should handle empty children', () => {
      const { container } = renderWithoutRouter(<Button />);

      expect(container.querySelector('button')).toBeInTheDocument();
    });

    it('should handle complex children (with icons)', () => {
      const { getByRole, getByText } = renderWithoutRouter(
        <Button>
          <span className="icon">🔍</span>
          <span>Search</span>
        </Button>
      );

      expect(getByRole('button')).toBeInTheDocument();
      expect(getByText('Search')).toBeInTheDocument();
    });

    it('should handle very long text', () => {
      const longText = 'This is a very long button text that might overflow';
      const { getByRole } = renderWithoutRouter(<Button>{longText}</Button>);

      expect(getByRole('button')).toHaveTextContent(longText);
    });

    it('should apply multiple classes correctly', () => {
      const { getByRole } = renderWithoutRouter(
        <Button variant="danger" size="lg" fullWidth loading className="custom">
          Delete
        </Button>
      );

      const button = getByRole('button');
      expect(button).toHaveClass('btn');
      expect(button).toHaveClass('btn-danger');
      expect(button).toHaveClass('btn-lg');
      expect(button).toHaveClass('btn-full-width');
      expect(button).toHaveClass('btn-loading');
      expect(button).toHaveClass('custom');
    });
  });

  describe('Snapshot Tests', () => {
    it('should match snapshot for default button', () => {
      const { container } = renderWithoutRouter(<Button>Click me</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for loading button', () => {
      const { container } = renderWithoutRouter(<Button loading>Loading</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for disabled button', () => {
      const { container } = renderWithoutRouter(<Button disabled>Disabled</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
