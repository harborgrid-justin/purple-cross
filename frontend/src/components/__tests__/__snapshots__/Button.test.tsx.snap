// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshot Tests > should match snapshot for default button 1`] = `
<button
  class="btn btn-primary btn-md"
>
  Click me
</button>
`;

exports[`Button Component > Snapshot Tests > should match snapshot for disabled button 1`] = `
<button
  class="btn btn-primary btn-md"
  disabled=""
>
  Disabled
</button>
`;

exports[`Button Component > Snapshot Tests > should match snapshot for loading button 1`] = `
<button
  class="btn btn-primary btn-md btn-loading"
  disabled=""
>
  <span
    aria-hidden="true"
    class="spinner"
  />
  <span>
    Loading...
  </span>
</button>
`;
